File: backend/src/github/store.ts
import { prisma } from "../prisma.js";
import { encryptToken, decryptToken } from "./crypto.js";

export async function saveUserToken(input: {
  workspaceId: string; githubUserId: number; login: string; avatarUrl: string; accessToken: string; scope: string;
}): Promise<void> {
  const data = {
    login: input.login, avatarUrl: input.avatarUrl,
    accessToken: encryptToken(input.accessToken), scope: input.scope,
  };
  await prisma.githubUserToken.upsert({
    where: { workspaceId_githubUserId: { workspaceId: input.workspaceId, githubUserId: input.githubUserId } },
    create: { workspaceId: input.workspaceId, githubUserId: input.githubUserId, ...data },
    update: data,
  });
}

export async function getUserToken(workspaceId: string): Promise<{ login: string; avatarUrl: string; accessToken: string } | null> {
  const row = await prisma.githubUserToken.findFirst({ where: { workspaceId }, orderBy: { createdAt: "asc" } });
  if (!row) return null;
  return { login: row.login, avatarUrl: row.avatarUrl, accessToken: decryptToken(row.accessToken) };
}

export async function deleteUserToken(workspaceId: string): Promise<void> {
  await prisma.githubUserToken.deleteMany({ where: { workspaceId } });
}

export async function saveInstallation(i: {
  workspaceId: string; installationId: number; accountLogin: string; accountType: string; repositorySelection: string;
}): Promise<void> {
  const data = { accountLogin: i.accountLogin, accountType: i.accountType, repositorySelection: i.repositorySelection };
  await prisma.githubInstallation.upsert({
    where: { installationId: i.installationId },
    create: { installationId: i.installationId, workspaceId: i.workspaceId, ...data },
    update: { ...data, workspaceId: i.workspaceId },
  });
}

export async function listInstallations(workspaceId: string): Promise<Array<{
  installationId: number; accountLogin: string; accountType: string; repositorySelection: string;
}>> {
  const rows = await prisma.githubInstallation.findMany({ where: { workspaceId }, orderBy: { createdAt: "asc" } });
  return rows.map((r) => ({
    installationId: r.installationId, accountLogin: r.accountLogin,
    accountType: r.accountType, repositorySelection: r.repositorySelection,
  }));
}

File: backend-py/app/github/store.py
import sqlalchemy as sa
from sqlalchemy.ext.asyncio import AsyncSession

from ..ids import new_id
from ..models import GithubInstallation, GithubUserToken
from .crypto import decrypt_token, encrypt_token


async def save_user_token(
    db: AsyncSession, *, workspace_id: str, github_user_id: int, login: str,
    avatar_url: str, access_token: str, scope: str,
) -> None:
    row = (
        await db.execute(
            sa.select(GithubUserToken).where(
                GithubUserToken.workspace_id == workspace_id,
                GithubUserToken.github_user_id == github_user_id,
            )
        )
    ).scalar_one_or_none()
    enc = encrypt_token(access_token)
    if row is None:
        db.add(GithubUserToken(
            id=new_id(), workspace_id=workspace_id, github_user_id=github_user_id,
            login=login, avatar_url=avatar_url, access_token=enc, scope=scope,
        ))
    else:
        row.login = login
        row.avatar_url = avatar_url
        row.access_token = enc
        row.scope = scope
    await db.commit()


async def get_user_token(db: AsyncSession, workspace_id: str) -> dict | None:
    row = (
        await db.execute(
            sa.select(GithubUserToken)
            .where(GithubUserToken.workspace_id == workspace_id)
            .order_by(GithubUserToken.created_at.asc())
        )
    ).scalars().first()
    if row is None:
        return None
    return {"login": row.login, "avatarUrl": row.avatar_url, "accessToken": decrypt_token(row.access_token)}


async def delete_user_token(db: AsyncSession, workspace_id: str) -> None:
    await db.execute(sa.delete(GithubUserToken).where(GithubUserToken.workspace_id == workspace_id))
    await db.commit()


async def save_installation(
    db: AsyncSession, *, workspace_id: str, installation_id: int, account_login: str,
    account_type: str, repository_selection: str,
) -> None:
    row = (
        await db.execute(
            sa.select(GithubInstallation).where(GithubInstallation.installation_id == installation_id)
        )
    ).scalar_one_or_none()
    if row is None:
        db.add(GithubInstallation(
            id=new_id(), installation_id=installation_id, workspace_id=workspace_id,
            account_login=account_login, account_type=account_type, repository_selection=repository_selection,
        ))
    else:
        row.workspace_id = workspace_id
        row.account_login = account_login
        row.account_type = account_type
        row.repository_selection = repository_selection
    await db.commit()


async def list_installations(db: AsyncSession, workspace_id: str) -> list[dict]:
    rows = (
        await db.execute(
            sa.select(GithubInstallation)
            .where(GithubInstallation.workspace_id == workspace_id)
            .order_by(GithubInstallation.created_at.asc())
        )
    ).scalars().all()
    return [
        {
            "installationId": r.installation_id, "accountLogin": r.account_login,
            "accountType": r.account_type, "repositorySelection": r.repository_selection,
        }
        for r in rows
    ]
